desktop/notification/fdo.go

```
@@ -0,0 +1,181 @@
+	}
+	return caps, nil
+}
+
+// SendNotification sends a single notification message.
+func (srv *fdoServer) SendNotification(msg *Message) (ID, error) {
+	call := srv.obj.Call(dBusInterfaceName+".Notify", 0,
+		msg.AppName, msg.ReplacesID, msg.Icon, msg.Summary, msg.Body,
+		flattenActions(msg.Actions), mapHints(msg.Hints), msg.ExpireTimeout)
+	var id ID
+	if err := call.Store(&id); err != nil {
+		return 0, err
+	}
+	return id, nil
+}
+
+func flattenActions(actions []Action) []string {
+	result := make([]string, len(actions)*2)
+	for i, action := range actions {
+		result[i*2] = action.ActionKey
+		result[i*2+1] = action.LocalizedText
+	}
+	return result
+}
+
+func mapHints(hints []Hint) map[string]dbus.Variant {
+	result := make(map[string]dbus.Variant, len(hints))
+	for _, hint := range hints {
+		result[hint.Name] = dbus.MakeVariant(hint.Value)
+	}
+	return result
+}
+
+// CloseNotification closes a specific notification.
+func (srv *fdoServer) CloseNotification(id ID) error {
+	call := srv.obj.Call(dBusInterfaceName+".CloseNotification", 0, id)
+	return call.Store()
+}
+
+// ObserveNotifications blocks and processes message notification signals.
+//
+// The bus connection is configured to deliver signals from the notification
+// server. All received signals are dispatched to the provided observer. This
+// process continues until stopped by the context, or if an error occurs.
+func (srv *fdoServer) ObserveNotifications(ctx context.Context, observer Observer) (err error) {
+	// TODO: upgrade godbus and use un-buffered channel.
+	ch := make(chan *dbus.Signal, 10)
+	defer close(ch)
+
+	srv.conn.Signal(ch)
+	defer srv.conn.RemoveSignal(ch)
+
+	matchRules := []dbus.MatchOption{
+		dbus.WithMatchSender(dBusName),
+		dbus.WithMatchObjectPath(dBusObjectPath),
+		dbus.WithMatchInterface(dBusInterfaceName),
+	}
+	if err := srv.conn.AddMatchSignal(matchRules...); err != nil {
+		return err
+	}
+	defer func() {
+		err = srv.conn.RemoveMatchSignal(matchRules...)
```

Will this clobber pre-exiting errors (e.g. from failures in processSignal() below)?

---

I think so, I wonder if just ignoring this error and surfacing whatever was provided is a better idea.